### PR TITLE
External MCP Tools Integration with Claude Desktop

### PR DESCRIPTION
**Pull request by anthonyjj89**
-----
**Note from me** 

This is my first ever GitHub contribution and I've only been coding for about 6 months, so I'm open to any feedback or suggestions for improvement. Sorry in advance for anything Ive done wrong but ive followed your guidelines and tutorials on this so hope its all ok :)

- This integration has been specifically tested and verified with Claude Desktop
- The implementation uses the [cline-mcp-tools](https://github.com/anthonyjj89/cline-mcp-tools) MCP server (which is still receiving improvements and updates but is useable in the moment of this PR submission)
- Future enhancements will extend support to other external MCP tools beyond Claude Desktop
- I'm excited to contribute this feature as I believe it creates a valuable workflow for users who want to leverage both Cline and Claude Desktop

--------------------------------

This PR introduces two key features that enable seamless integration between Cline and external MCP tools like Claude Desktop:

- Active Conversation Marking: Users can mark conversations as "Active A" or "Active B", allowing external MCP tools to detect and access these conversations.
- Notification System Bridge: External MCP tools can send messages, suggestions, and advice to active Cline conversations through the notification system.

**This creates a cost-effective workflow where users can:**

- Use Claude Desktop for research and planning (with its $20/month subscription)
- Send targeted advice and solutions to Cline through the notification system
- Leverage Cline's powerful code editing capabilities without incurring per-message costs for research and planning
- Test Procedure

I've manually tested these features by:

- Verifying notification creation and display
- Testing notification interaction (read, dismiss, restore, send to chat)
- Testing active conversation toggling
- Verifying integration with Claude Desktop via the cline-mcp-tools server
- Testing edge cases (e.g., no active task, multiple extension instances)
- All tests passed successfully and the features work as expected.

## MCP Tools Installation

To use this feature, users need to install the cline-mcp-tools MCP server:

### Prerequisites
- Node.js 18.0.0 or higher
- Claude Desktop application
- **IMPORTANT**: The forked version of the Cline VS Code extension from: https://github.com/anthonyjj89/cline-mcp-external-link (this one)

### Download and Build
1. Clone the repository:
   ```bash
   git clone https://github.com/anthonyjj89/cline-mcp-tools.git
   cd cline-mcp-tools

For detailed installation and configuration instructions, see the [cline-mcp-tools README](https://github.com/anthonyjj89/cline-mcp-tools#readme).

**Quick Setup of said MCP Server for Claude Desktop**
Configure Claude Desktop to use this MCP server by adding the following to your Claude Desktop configuration file:

```
"mcpServers": {
  "cline-chat-reader": {
    "command": "node",
    "args": [
      "/path/to/cline-mcp-tools/build/index.js"
    ],
    "disabled": false,
    "autoApprove": []
  }
}
```
Replace /path/to/cline-mcp-tools with the actual path where you cloned the repository.

The send_external_advice tool will become available in Claude Desktop automatically after restarting.

------------------

Type of Change
[ ] 🐛 Bug fix (non-breaking change which fixes an issue)
[x] ✨ New feature (non-breaking change which adds functionality)
[ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
[x] 📚 Documentation update
Pre-flight Checklist
[x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
[x] Tests are passing (npm test) and code is formatted and linted (npm run format && npm run lint)
[x] I have created a changeset using npm run changeset (required for user-facing changes)
[x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

**Screenshots**

<img width="726" alt="image" src="https://github.com/user-attachments/assets/a1a36930-1678-4b70-be46-dc6e74deace2" />

<img width="724" alt="image" src="https://github.com/user-attachments/assets/cac4cc33-21ff-43c8-9be9-4c91125d9f62" />

<img width="836" alt="image" src="https://github.com/user-attachments/assets/4bdb9efe-3d67-4c49-8f9b-8ecc0caac068" />

<img width="874" alt="image" src="https://github.com/user-attachments/assets/b062f1c7-d8b6-45ba-a9fc-04b9dd349860" />

<img width="1112" alt="image" src="https://github.com/user-attachments/assets/1331e034-f0e7-45e5-87e7-e65e7f43e0b1" />

<img width="770" alt="image" src="https://github.com/user-attachments/assets/c1c716b9-5840-449d-831c-1b4f9bc18b18" />


**Additional Notes**

- This integration has been specifically tested and verified with Claude Desktop
- The implementation uses the [cline-mcp-tools](https://github.com/anthonyjj89/cline-mcp-tools) MCP server
- Future enhancements will extend support to other external MCP tools beyond Claude Desktop
<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Integrates Cline with external MCP tools like Claude Desktop by adding active conversation marking and a notification system.
> 
>   - **Integration Features**:
>     - Adds active conversation marking in `ClineProvider.ts` to toggle conversations as "Active A" or "Active B".
>     - Implements notification system in `ClineProvider.ts` to handle external advice from MCP tools.
>   - **Notification System**:
>     - Adds `sendExternalAdvice()`, `markAdviceAsRead()`, `dismissAdvice()`, and `restoreAdvice()` in `ClineProvider.ts`.
>     - Updates `ExtensionMessage.ts` and `WebviewMessage.ts` to handle new message types for notifications.
>   - **UI Components**:
>     - Adds `NotificationPanel` component in `NotificationPanel.tsx` to display notifications.
>     - Updates `TaskHeader.tsx` to include active conversation toggle and notification bell.
>   - **Documentation**:
>     - Updates `active-conversation.md` and `notifications.md` to document new features.
>   - **Miscellaneous**:
>     - Adds `compare-files.sh` script for comparing file changes.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 8388bb91502232860847a7bd6f7d694b81a82ee7. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

### CODE REVIEW

**Comment by hackertron** [comment]
"IMPORTANT: The forked version of the Cline VS Code extension from: https://github.com/anthonyjj89/cline-mcp-external-link (this one)". this won't be needed once it's merged right ?

**Comment by anthonyjj89** [comment]
> "IMPORTANT: The forked version of the Cline VS Code extension from: https://github.com/anthonyjj89/cline-mcp-external-link (this one)". this won't be needed once it's merged right ? 

Yeah that's right. If the team decides it's a useful feature to have in the main cline then you'll just need the MCP server I made. But until then you need the MCP server and this modified version of cline. 

Hope that makes sense. Sorry if it wasn't clear.

File: CHANGELOG.md
Comment (dcbartlett) [inline CHANGELOG.md:1]: Please don't update the ChangeLog.  We use changesets to automatically update this.  Please remove CHANGELOG.md from the PR.

File: .changeset/rotten-gifts-jam.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
...
...n features for external MCP tools integrationThis changeset introduces two key features that enable seamless integration between Cline and external MCP tools like Claude Desktop:1. Notification System Bridge: - Allows external MCP tools to send messages, suggestions, and advice to active Cline conversations - Implements a notification panel with read, dismiss, and restore functionality - Provides a "Send to Chat" feature to easily incorporate external advice into conversations2. Active Conversation Marking: - Enables users to mark conversations as "Active A" or "Active B" - Allows external MCP tools to detect and access these conversations - Creates a seamless workflow between Cline and other AI assistantsThese features are implemented using the cline-mcp-tools MCP server (https://github.com/anthonyjj89/cline-mcp-tools) and have been tested with Claude Desktop.This integration creates a cost-effective workflow where users can use Claude Desktop for research and planning, then send targeted advice to Cline without incurring per-message costs.Note: Currently, the integration has only been tested with Claude Desktop. Future enhancements will extend support to other external MCP tools.
```
Comment (dcbartlett) [inline .changeset/rotten-gifts-jam.md]: This is going to be a VERY long entry for our changeset automation.  Can you summarize this to about 1 or 2 sentences?

File: ~/cline-pr-prep/compare-files.sh
Comment (dcbartlett) [inline ~/cline-pr-prep/compare-files.sh:1]: I haven't completely looked through all of the files here yet, but can you explain what this file is used for?

**Review by dcbartlett — CHANGES_REQUESTED** [review]

File: ~/cline-pr-prep/compare-files.sh
Comment (anthonyjj89) [reply ~/cline-pr-prep/compare-files.sh:1]: Hi @dcbartlett 

Thank you for the feedback! I've restored the CHANGELOG.md file from the main branch, shortened the changeset message to be more concise as requested.

Regarding compare-files.sh, I apologize for including this - it was part of my PR preparation process and wasn't meant to be submitted. I've removed the entire pr-prep folder from the PR.

**Comment by anthonyjj89** [comment]
Hey team! 👋 I've updated the Dev selector with some nice improvements:

- Replaced the toggle with a dropdown menu showing all options at once
- Added color-coded status indicators to show which are already assigned in other instances of Cline (green=available, red=in-use)
- Created a confirmation dialog to prevent accidental reassignments
- Built more robust selection logic using external JSON storage that works across multiple VS Code windows and extension instances
- Each Dev (A-D) can now be tracked and communicated with independently

These changes make the selection process more intuitive while enabling seamless integration with external tools. 

What do you think? 😃 

Xnapper-2025-03-28-22 23 04
Xnapper-2025-03-28-22 23 32
Xnapper-2025-03-28-22 24 11

**Review by celestial-vault — CHANGES_REQUESTED** [review]
Hey Anthony,

I appreciate the effort that you put into your PR description and the thought that you evidently put into this PR. Welcome to coding and the world of open source!

That being said, could you please break this PR into bite sized pieces? Reviewing 2000+ lines of code is no small feat :face_holding_back_tears:

The ideal flow is that we can look at your PR and see what one single thing it's doing very quickly. Something useful to look up is how to open branches on other branches for easier encapsulation.

Thanks

**Comment by celestial-vault** [comment]
> > Hey Anthony,
> > I appreciate the effort that you put into your PR description and the thought that you evidently put into this PR. Welcome to coding and the world of open source!
> > That being said, could you please break this PR into bite sized pieces? Reviewing 2000+ lines of code is no small feat 🥹
> > The ideal flow is that we can look at your PR and see what one single thing it's doing very quickly. Something useful to look up is how to open branches on other branches for easier encapsulation.
> > Thanks
> 
> Hey @celestial-vault!
> 
> I really appreciate the kind words. I was very worried the correct procedures and such for contributing. Quite nervous to hit the submit button.
> 
> I understand what you are asking completely. If you would be so awesome to explain a good way to break it up into more manageable chunks as all these changes sort of work in tandem to provide the feature.
> 
> I can break the PRs into smaller changes to each component/file but need to research how this open branches onto other branches works. I will 100% check that out today or tomorrow. Thanks for the tip.
> 
> I will try to isolate each part into maybe 600 lines per PR perhaps? Apologies for being unfamiliar with this. Just curious how you check only part of a code change without seeing the actual results of it all working together :)
> 
> Another thing I considered is to make a quick YT video to include in the PR basically explaining the feature and how it works under the hood.
> 
> You guys might be thrilled to hear that along with lots of learning of structure of the code, a lot of this was in fact written with cline itself which I find quite wonderful. What a tool!
> 
> Please let me know if you would prefer me to try break it up into more manageable chunks or if indeed a video explanation would help keeping our current PR.
> 
> Thank you very much for the consideration. I can tell you I'm using this version on my own work now and it really is very helpful.
> 
> Thank you again for being kind! 😃

@anthonyjj89 

Hey, yeah of course.

Basically, you can do `git checkout -b "your-new-branch"` while on a branch to make a branch OFF of that branch. Then, when you open a PR, you can choose the base to be the branch that you made the PR off of, rather than main. Ask ChatGPT about it and you'll quickly find out.

As far as splitting up a PR, yes, this is a tricky thing to do. Basically, try to do things in isolated units. For a feature this large, I would make sure you get buy-in in the discussions tab or the Discord before spending a lot of time on it.

In this particular PR, it seems you are adding many complementary features. I would think hard about how they could stand on their own.

**Comment by celestial-vault** [comment]
Going to close this for now. Seems like an interesting idea, just hit us back up with something a bit more bitesized. Think incremental improvements.

**Comment by anthonyjj89** [comment]
@celestial-vault sorry for not replying for a couple days. I understand these things move quickly as well is what I've been told!

No problem at all! I really appreciate the information/education on it all and I have some smaller ideas to submit instead. I might turn this functionality into just an MCP server to bridge the connection between claude desktop and cline that way for now and revisit this as a built in feature later on. Might be easier. 
Do you have a certain place I can submit MCPs for the marketplace within cline?

Thanks again Evan for being accommodating to us beginners. 😃 
Anthony
